src/scope.litcoffee

```diff
@@ -94,27 +99,61 @@ compiler-generated variable. `_var`, `_var2`, and so on...
           break unless @check(temp) or temp in @root.referencedVars
           index++
         @add temp, 'var', yes if options.reserve ? true
+        @laterVar temp if options.laterVar
         temp
 
-Ensure that an assignment is made at the top of this scope
-(or at the top-level scope, if requested).
+Ensure that an assignment is made at the top of this scope.
 
       assign: (name, value) ->
-        @add name, {value, assigned: yes}, yes
-        @hasAssignments = yes
+        @get(name).assigned = value
+
+Add a comment that should appear when the variable is declared
+(for Flow support).
+
+      comment: (name, comments) ->
+        @get(name).comments = comments
+
+Does this variable have a comment attached to it in this scope?
+
+      hasComment: (name) ->
+        @get(name)?.comments?
+
+Check whether a var declaration of this variable could go later, and if so,
+mark it as so.
+
+      laterVar: (name) ->
+        # Ensure variable is declared at this scope, as a regular 'var', and
+        # we haven't already given it a var prefix somewhere, and it doesn't
+        # have an attachment that goes with a top var declaration.
+        v = @get name
+        later = v?.type is 'var' and
+                not (v.laterVar or v.comments?)
+        v.laterVar = yes if later
+        later
 
 Does this scope have any declared variables?
 
       hasDeclarations: ->
-        !!@declaredVariables().length
+        return true for v in @variables when v.type is 'var' and not v.laterVar
+
+Return a list of names of variables declared in this scope.
+Optionally restrict to assigned or unassigned variables.
+
+      declaredVariables: (assigned) ->
+        (v.name for v in @variables when v.type is 'var' and not v.laterVar and
+          switch assigned
+            when true then v.assigned?
+            when false then not v.assigned?
+            else true
+        ).sort()
```

Comment:
Isn’t that switch just `assigned == v.assigned?`?

Reply:
The `switch` deals with three cases: `assigned` can be

* `true` meaning "only give me assigned variables" i.e. `when v.assigned?`
* `false` meaning "only give me unassigned variables" i.e. `when not v.assigned?` 
* `undefined` (or absent argument) meaning "give me all variables" i.e. `when true`.

I forget whether the `undefined` case is actually used; if it isn't, we could definitely simplify. But I don't think either simplification above is correct according to the 3-case spec.

Thanks @GeoffreyBooth for the thorough review! I will work on responding and adding a bunch of comments.

Reply:
It might be going a bit far to do this all in a one-liner. You could maybe have a one-liner that retrieves all variables, then something like “if give me only assigned variables, filter the collection to just those; else if only give me unassigned variables, filter the collection to just those.” Then sort whatever’s left in the collection at this point and return it. Making this a multiline function also gives you opportunities to add comments explaining each branch and why you might want each.